### PR TITLE
spotify browse webapi: passthrough implementation, no cache

### PR DESCRIPTION
**Pull request by fredericgermain**
https://developer.spotify.com/web-api/browse-endpoints/

bugs to fix (patches are coming soon):
- refresh token
- maybe prefix these urls with browse: ?
- multi-page

### CODE REVIEW

**Comment by fredericgermain** [comment]
also missing, implementation for images...

File: mopidy_spotify_web/library.py [hunk 1]
```diff
@@ -97,6 +100,28 @@ def browse(self, uri):
         elif uri == 'spotifyweb:albums':
             return self._cache.sortedAlbums
             # return Ref directory for all albums
+        elif uri.startswith('spotifyweb:featured-playlists') or \
```
Comment (adamcik) [inline mopidy_spotify_web/library.py:125]: `uri.startswith(('spotifyweb:featured-playlists', 'spotifyweb:new-releases', 'spotifyweb:categories'))` works btw :-)

Personally I also like to replace this with something like https://github.com/mopidy/mopidy-dirble/blob/master/mopidy_dirble/translator.py#L14-18 so you just the the variant and identifier back to work with.

File: mopidy_spotify_web/library.py [hunk 2]
```diff
@@ -97,6 +100,28 @@ def browse(self, uri):
...
+                    arr += [ Ref.directory(uri='spotifyweb:categories:'+cat['id']+':playlists',
+                                    name=cat['name']) for cat in res['categories']['items'] ]
+                elif res.has_key('playlists'):
+                    arr += [ Ref.playlist(uri=playlist['uri'],
+                                    name=playlist['name']) for playlist in res['playlists']['items'] ]
+                elif res.has_key('albums'):
+                    arr += [ Ref.album(uri=album['uri'],
+                                    name=album['name']) for album in res['albums']['items'] ]
+                return arr
+            except Exception as e:
```
Comment (adamcik) [inline mopidy_spotify_web/library.py]: Please catch more specific errors than this. Likely you want to only wrap the spotipy call. Also use `logger.error(...)`

**Comment by adamcik** [comment]
Cool, was hoping to get around to adding something like this myself. As for the refresh token #2 will give some slight changes there, but if all you plan to do is call the get token function it shouldn't effect what you have planned.

**Comment by fredericgermain** [comment]
I pushed a code that refresh the token based on current time and the expires_in field from the /refresh call. I do that when it's supposed to be about to exire, minus 60 secs

changed the execption handling

manage multi-page request, basically using the code from get_tracks_from_web_api()

I'm still undecided on how decide the final "spotifyweb:" uri of these new objects...
@adamcik , using a (variant, id) might be a little weird, as there is no 'id' on the featured-playlist or new-features pages... and the id for a category would be just a name, like 'toplists', 'mood'

I'd better prefix these url with 'browse', as we'd get a 1:1 mapping with spotify api : 'spotifyweb:browse:categories:toplists' -> .../browse/categories/toplists'

Well, let me know what you think

**Comment by adamcik** [comment]
As long as the first part is something that correctly parses to the scheme `spotifyweb` what the rest does is really up to what makes sense for the extension. So picking something that is reasonable to work with and maps nicely to the API makes sense.

As for use of the URI, only the backend should actually care about what it is, the rest of the system (other backends, frontends, etc) should treat the URIs as opaque strings, the exception being the mopidy core which is allowed to look at the URI to route things to the right backend.

**Pull request merged**